# **Support Vector Regression (SVR) Project**
# Medical Cost Prediction & Feature Impact Analysis

---

## **Introduction**

This project focuses on building a deep understanding of how machine learning models can be used to model real world problems, specifically in the healthcare domain. Using the Medical Cost Personal Dataset, the goal is not only to predict insurance charges but also to analyze how different features influence model behavior and performance.

Rather than treating this as a simple regression task, this project is approached as a full machine learning experiment, where different modeling choices, preprocessing techniques, and assumptions are carefully tested and evaluated.

---

## **Project Aim**

The aim of this project is to develop, analyze, and compare multiple regression models to accurately predict medical insurance costs while gaining a strong conceptual and practical understanding of machine learning techniques.

---

## **Objectives**

1. Build a solid understanding of regression algorithms through hands-on implementation
2. Perform in-depth exploratory data analysis (EDA) to uncover patterns and relationships
3. Analyze the effect of key features (especially BMI and smoking) on medical costs
4. Apply and compare different preprocessing techniques:
    - Encoding categorical variables
    - Feature scaling
    - Log transformation (for skewed data)
5. Train and evaluate multiple models, such as:
    - Linear Regression
    - Polynomial Regression
    - Support Vector Regression (SVR)
    - Decision Tree
    - Random Forest
6. Perform model comparison and critical analysis, not just reporting scores
7. Understand why a model performs well or fails, not just the final result

---
## **Dataset Description**

The dataset includes the following features:

- age → Age of the individual
- sex → Gender
- bmi → Body Mass Index
- children → Number of dependents
- smoker → Smoking status
- region → Residential area
- charges → Medical insurance cost (target variable)

The dataset used in this project is available on Kaggle: [Dataset Link](https://www.kaggle.com/datasets/d3lhomi10/medical-cost-personal-dataset)

---

## **Problem Type**

This is a Supervised Machine Learning – Regression Problem, where the goal is to predict a continuous target variable (charges).

---

Project Approach

This project is structured as a complete ML pipeline, including:

1. Data Exploration (EDA)
2. Data Preprocessing
3. Feature Engineering
4. Model Training
5. Model Evaluation
6. Model Comparison
7. Insights & Conclusions

The focus is not only on building a working model, but on understanding each step deeply and justifying every decision.




# **Importing the Libraries**

In [2]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# **Data preprocessing**

## **Data Cleaning Check Template**
This template is designed to quickly assess the quality of any dataset before building machine learning models or performing analysis.

It provides a structured overview of the dataset by checking for common data issues such as:

- Missing values

- Duplicate rows

- Incorrect data types

- Outliers

- Distribution of numerical features

- Categorical feature consistency

**What This Template Does**

- Displays basic dataset information (shape, data types)

- Identifies missing values and duplicates

- Summarizes numerical and categorical features

- Detects potential outliers using the IQR method

- Highlights columns with low unique values for quick inspection

How to Use

1. Load your dataset using Pandas  

2. Call the function:

In [3]:
def data_quality_report(df):

    print("DATA QUALITY REPORT")
    
    # Print a separator line for better readability
    
    print("=" * 50)
    print("BASIC INFO")
    print("=" * 50)
    
    # Show general information about the dataset (columns, data types, non-null values)
    print(df.info())
    
    # Show number of rows and columns
    print("\n" + "=" * 50)
    print("SHAPE OF DATA")
    print("=" * 50)
    print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
    
    # Check for missing (null) values in each column
    print("\n" + "=" * 50)
    print("MISSING VALUES")
    print("=" * 50)
    missing = df.isnull().sum()
    
    # Display only columns that have missing values
    print(missing[missing > 0])
    
    # Check for duplicate rows
    print("\n" + "=" * 50)
    print("DUPLICATES")
    print("=" * 50)
    print(f"Duplicate rows: {df.duplicated().sum()}")
    
    # Display data types of each column
    print("\n" + "=" * 50)
    print("DATA TYPES")
    print("=" * 50)
    print(df.dtypes)
    
    # Summary statistics for numerical columns (mean, std, min, max, etc.)
    print("\n" + "=" * 50)
    print("NUMERICAL SUMMARY")
    print("=" * 50)
    print(df.describe())
    
    # Summary for categorical (object) columns
    print("\n" + "=" * 50)
    print("CATEGORICAL SUMMARY")
    print("=" * 50)
    print(df.describe(include=['object']))
    
    # Show unique values for columns with low number of distinct values
    # Useful for detecting categories, errors, or inconsistencies
    print("\n" + "=" * 50)
    print("UNIQUE VALUES (LOW CARDINALITY)")
    print("=" * 50)
    for col in df.columns:
        if df[col].nunique() < 10:  # Only show columns with few unique values
            print(f"{col}: {df[col].unique()}")
            
    # correlation
    print("\n" + "=" * 50)
    print("CORRELATION MATRIX")
    print("=" * 50)
    print(df.corr(numeric_only=True))
    
    # Detect outliers using the IQR (Interquartile Range) method
    print("\n" + "=" * 50)
    print("OUTLIERS CHECK (IQR METHOD)")
    print("=" * 50)
    
    # Loop through only numerical columns
    for col in df.select_dtypes(include=np.number).columns:
        Q1 = df[col].quantile(0.25)  # 25th percentile
        Q3 = df[col].quantile(0.75)  # 75th percentile
        IQR = Q3 - Q1  # Interquartile range
        
        # Count rows that fall outside the normal range
        outliers = df[(df[col] < Q1 - 1.5 * IQR) | (df[col] > Q3 + 1.5 * IQR)]
        print(f"{col}: {len(outliers)} outliers")

# **Load dataset**
Apply Data Cleaning Check Template

In [ ]:
dataset = pd.read_csv("insurance.csv")
data_quality_report(dataset)

DATA QUALITY REPORT
BASIC INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB
None

SHAPE OF DATA
Rows: 1338, Columns: 7

MISSING VALUES
Series([], dtype: int64)

DUPLICATES
Duplicate rows: 1

DATA TYPES
age           int64
sex          object
bmi         float64
children      int64
smoker       object
region       object
charges     float64
dtype: object

NUMERICAL SUMMARY
               age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397 

# **Cleaning the data** 

### remove the duplication

In [5]:
# Step 2: Remove duplicates FROM the cleaned dataset
dataset_clean = dataset.drop_duplicates().copy()

# Final result
print(dataset_clean)

      age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1337 rows x 7 columns]


# **Select Independent Variable (X) and Dependent Variable (y)**

Note: The concepts of Independent and Dependent variables were discussed earlier.

If you would like to review or gain a clearer understanding, please refer to this section: [Go to Independent variable VS Dependent variable Section](https://github.com/Hazem1695/ml-concept-briefs)

In [6]:
# Separate independent variable(features (X)) by selecting all rows and all columns except the last one
X = dataset_clean.iloc[:, :-1].values
# Separate the dependent variable (Target(y)) by selecting the last column
y = dataset_clean.iloc[:, -1].values

In [7]:
# reshape y before to apply feature scaling
y = y.reshape(len(y), 1)

In [8]:
y_log = np.log(y)
print(y_log)

[[ 9.73417643]
 [ 7.45330245]
 [ 8.40053847]
 ...
 [ 7.39623314]
 [ 7.60486709]
 [10.27991376]]


In [9]:
print(X)

[[19 'female' 27.9 0 'yes' 'southwest']
 [18 'male' 33.77 1 'no' 'southeast']
 [28 'male' 33.0 3 'no' 'southeast']
 ...
 [18 'female' 36.85 0 'no' 'southeast']
 [21 'female' 25.8 0 'no' 'southwest']
 [61 'female' 29.07 0 'yes' 'northwest']]


In [10]:
print(y)

[[16884.924 ]
 [ 1725.5523]
 [ 4449.462 ]
 ...
 [ 1629.8335]
 [ 2007.945 ]
 [29141.3603]]


# **Encoding categorical data**

## Encoding the Indpendent Variable Using One Hot Encoding

Note: I explained One-Hot Encoding earlier.

If you missed it or want a deeper understanding of how it works, you can find it here: [Go to One-Hot Encoding Section](https://github.com/Hazem1695/ml-concept-briefs)

In [11]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
# One-hot encode the 'color' column (index 2) while keeping other features
# Initialize ColumnTransformer to handle categorical data
# [2] targets the 'color' column (index 2)
ct = ColumnTransformer(transformers = [('encoder', OneHotEncoder(sparse_output=False),[1,4,5])], remainder = 'passthrough')
# Execute transformation and convert to NumPy ndarray for model training
X = np.array(ct.fit_transform(X))

In [12]:
print(X)

[[1.0 0.0 0.0 ... 19 27.9 0]
 [0.0 1.0 1.0 ... 18 33.77 1]
 [0.0 1.0 1.0 ... 28 33.0 3]
 ...
 [1.0 0.0 1.0 ... 18 36.85 0]
 [1.0 0.0 1.0 ... 21 25.8 0]
 [1.0 0.0 0.0 ... 61 29.07 0]]


# **Splitting the dataset into the Train set and Test set**

In [13]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train , y_test = train_test_split(X, y_log , test_size = 0.2, random_state = 0)

In [14]:
#np.set_printoptions(threshold=np.inf, linewidth=1000)
print(X_test)
#first column: [1.0 0.0 1.0 0.0 0.0 0.0 1.0 0.0 18 39.82 0]

[[1.0 0.0 1.0 ... 18 39.82 0]
 [1.0 0.0 1.0 ... 47 29.37 1]
 [0.0 1.0 1.0 ... 49 31.35 1]
 ...
 [0.0 1.0 1.0 ... 18 26.18 2]
 [0.0 1.0 0.0 ... 54 25.1 3]
 [1.0 0.0 1.0 ... 62 30.495 2]]


# **Feature Scaling**

Note: I explained Feature Scaling earlier.

If you want to understand when and why to use feature scaling, you can check it here: [Go to  When is Feature Scaling Necessary in Regression Models? Section](https://github.com/Hazem1695/ml-concept-briefs)

In [15]:
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
sc_y = StandardScaler()
X_train[:, [8, 9]] = sc_X.fit_transform(X_train[:,[8, 9]])
y_train = sc_y.fit_transform(y_train)
X_test[:, [8, 9]] = sc_X.transform(X_test[:, [8, 9]])

In [16]:
print(X_test)

[[1.0 0.0 1.0 ... -1.5104150060362007 1.4918145939943805 0]
 [1.0 0.0 1.0 ... 0.545901168138331 -0.23070850925496822 1]
 [0.0 1.0 1.0 ... 0.6877160767020918 0.09566428925543478 1]
 ...
 [0.0 1.0 1.0 ... -1.5104150060362007 -0.7565313512995064 2]
 [0.0 1.0 0.0 ... 1.0422533481114937 -0.934552877759726 3]
 [1.0 0.0 1.0 ... 1.609512982366537 -0.04526941919223928 2]]


In [17]:
print(f"Rows: {X_test.shape[0]}, Columns: {X_test.shape[1]}")

Rows: 268, Columns: 11


# **Building Support Vector Regression (SVR) Model**

## Training the Building Support Vector Regression (SVR) model on the Training set

In [18]:
from sklearn.svm import SVR
regressor = SVR(kernel = 'rbf')
regressor.fit(X_train, y_train.ravel())

SVR()

# **Preticting the Test set results**

In [19]:
y_pred_log = sc_y.inverse_transform(regressor.predict(X_test).reshape(-1,1))


y_pred = np.exp(y_pred_log)
np.set_printoptions(precision=2,suppress=True)
print(np.concatenate((y_pred.reshape(len(y_pred), 1), np.exp(y_test).reshape(len(y_test), 1)), axis = 1))

[[ 1829.08  1633.96]
 [ 9262.08  8547.69]
 [10118.94  9290.14]
 [19468.74 32548.34]
 [10146.54  9644.25]
 [ 2747.21  2680.95]
 [ 2390.23  2198.19]
 [ 1385.95  1241.56]
 [ 2847.06  2710.83]
 [13003.16 12235.84]
 [ 9049.24  8280.62]
 [17453.65 17043.34]
 [14828.21 13974.46]
 [ 8885.99  8219.2 ]
 [ 5816.78  5472.45]
 [ 3288.38  2438.06]
 [ 5585.74  5267.82]
 [ 3789.55  3490.55]
 [ 7155.73  6640.54]
 [15312.25 14692.67]
 [ 1759.91  1622.19]
 [14469.28 13224.69]
 [ 1410.4   1256.3 ]
 [ 2824.33  2643.27]
 [ 1604.11  1674.63]
 [ 4982.9   4667.61]
 [ 3884.57  3732.63]
 [11927.88 11552.9 ]
 [ 3839.84  3756.62]
 [25773.07 37465.34]
 [ 8580.98  8059.68]
 [44360.54 47462.89]
 [11334.85 10577.09]
 [11237.66 20630.28]
 [13208.01 14571.89]
 [15313.21 36580.28]
 [ 9076.76  8347.16]
 [44404.73 51194.56]
 [ 9461.77  8428.07]
 [ 2012.97  1880.49]
 [26657.28 33475.82]
 [ 2594.23  2867.12]
 [ 5114.08  4564.19]
 [50302.95 47496.49]
 [32851.31 36149.48]
 [ 8771.09  8125.78]
 [11707.81 19749.38]
 [ 8376.78  7

In [20]:
#y_pred = sc_y.inverse_transform(regressor.predict(X_test).reshape(-1,1))
#np.set_printoptions(precision=2)
#print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

# **Evaluating the Model Performance**

In [21]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred_log)

0.855586929659759

In [23]:
y_trained = regressor.predict(X_train)
r2_score(y_train, y_trained)

0.834508076984737


# Performance Summary Table

| Kernel / Method             | Test R² Score | Conclusion                                                                                         |
| --------------------------- | ------------- | -------------------------------------------------------------------------------------------------- |
| Linear                      | 0.7042        | Underfitting: unable to capture the non-linear structure in the data                               |
| Polynomial (Degree 2)       | 0.8347        | Strong performance: effectively captures key non-linear relationships                              |
| Polynomial (Degree 3)       | 0.8235        | Slight performance drop: slightly more complex, but less stable than degree 2                      |
| Polynomial (Degree 4)       | 0.7935        | Over-complexity reduces generalization performance                                                 |
| RBF (Radial Basis Function) | **0.8377**    | **Best kernel among standard SVR models: strong non-linear modeling ability and stability**        |
| Sigmoid                     | -254.12       | Model failure: incompatible with the data distribution and structure                               |
| Log Transformation          | **0.8556**    | **Best overall performance: improves stability and captures target distribution more effectively** |




# **Important Insights**

##### **1. Kernel Behavior Insight**

The results show that SVR performance is highly dependent on the kernel choice. Linear kernel underperforms due to its inability to model non-linear relationships in the data, while polynomial kernels demonstrate strong learning ability at lower degrees, especially degree 2.

##### **2. Effect of Increasing Polynomial Degree**

As the polynomial degree increases beyond 2, performance begins to decline. This indicates that higher complexity introduces noise sensitivity and reduces the model’s ability to generalize effectively.

##### **3. RBF Kernel Strength**

The RBF kernel achieves the highest performance among standard kernels, showing strong adaptability to non-linear patterns without excessive overfitting. This makes it the most reliable kernel choice for this dataset.

##### **4. Sigmoid Kernel Limitation**

The sigmoid kernel fails dramatically, indicating incompatibility with the underlying data distribution. This highlights that not all kernels are universally applicable, and performance is highly data dependent.

##### **5. Effect of Log Transformation**

Applying log transformation to the target variable significantly improves model performance, achieving the highest test R² (0.8556). This suggests that stabilizing the target distribution allows SVR to learn more meaningful patterns and improves predictive accuracy.

##### **6. Importance of feature and target scaling**
Feature scaling played a critical role in achieving strong performance. Since SVR relies on distance calculations, unscaled data can lead to poor optimization and unstable results. Scaling both features and the target variable ensured that the model could learn effectively and produce consistent predictions.

##### **7. Generalization performance**
The small gap between training and testing performance indicates that the model generalizes well to unseen data. This is a strong sign that the model is neither overfitting nor underfitting, but instead capturing the underlying structure of the dataset effectively.


# **Conclusion**

Overall, SVR performance is strongly influenced by kernel selection and data transformation. RBF and Polynomial (degree 2) provide the most reliable results, while log transformation further enhances performance by improving data stability. This confirms that both kernel choice and target distribution handling are critical factors in SVR modeling success.

# **Model Comparison**

| Model Type            | Configuration            | R² Score |
| --------------------- | ------------------------ | -------- |
| Linear Regression     | Baseline                 | 0.7530   |
| **Linear Regression**   | **+ Log Transformation**     | **0.7716** |
| Linear Regression     | + Log + Feature Scaling  | 0.7716   |
| Polynomial Regression | Baseline                 | 0.8349   |
| **Polynomial Regression** | **+ Log Transformation**   |  **0.8495**|
| Polynomial Regression | + Log + Feature Scaling  | 0.8495   |
| **SVR**                 | **RBF + Feature Scaling**   | **0.8377** |
| SVR                   | Polynomial (degree=2)    | 0.8347   |
| SVR                   | Polynomial (degree=3)    | 0.8235   |
| SVR                   | Polynomial (degree=4)    | 0.7935   |
| SVR                   | Linear Kernel            | 0.7043   |
| SVR                   | Sigmoid Kernel           | -254.12  |
| **SVR**                   | **RBF + Log Transformation** | **0.8556**   |



